drivers/sensor/lps22df/lps22df_trigger.c

```diff
@@ -0,0 +1,256 @@
+/* ST Microelectronics LPS22DF pressure and temperature sensor
+ *
+ * Copyright (c) 2023 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Datasheet:
+ * https://www.st.com/resource/en/datasheet/lps22df.pdf
+ */
+
+#define DT_DRV_COMPAT st_lps22df
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/logging/log.h>
+
+#include "lps22df.h"
+
+LOG_MODULE_DECLARE(LPS22DF, CONFIG_SENSOR_LOG_LEVEL);
+
+/**
+ * lps22df_enable_int - enable selected int pin to generate interrupt
+ */
+static int lps22df_enable_int(const struct device *dev, int enable)
+{
+	const struct lps22df_config * const cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lps22df_pin_int_route_t int_route;
+
+	/* set interrupt */
+	lps22df_pin_int_route_get(ctx, &int_route);
+	int_route.drdy_pres = enable;
+	return lps22df_pin_int_route_set(ctx, &int_route);
+}
+
+/**
+ * lps22df_trigger_set - link external trigger to event data ready
+ */
+int lps22df_trigger_set(const struct device *dev,
+			  const struct sensor_trigger *trig,
+			  sensor_trigger_handler_t handler)
+{
+	struct lps22df_data *lps22df = dev->data;
+	const struct lps22df_config * const cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lps22df_data_t raw_data;
+
+	if (trig->chan == SENSOR_CHAN_ALL) {
+		lps22df->handler_drdy = handler;
+		lps22df->data_ready_trigger = trig;
+		if (handler) {
+			/* dummy read: re-trigger interrupt */
+			if (lps22df_data_get(ctx, &raw_data) < 0) {
+				LOG_DBG("Failed to read sample");
+				return -EIO;
+			}
+			return lps22df_enable_int(dev, 1);
+		} else {
+			return lps22df_enable_int(dev, 0);
+		}
+	}
+
+	return -ENOTSUP;
+}
+
+/**
+ * lps22df_handle_interrupt - handle the drdy event
+ * read data and call handler if registered any
+ */
+static void lps22df_handle_interrupt(const struct device *dev)
+{
+	int ret;
+	struct lps22df_data *lps22df = dev->data;
+	const struct lps22df_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lps22df_all_sources_t status;
+
+	if (lps22df_all_sources_get(ctx, &status) < 0) {
+		LOG_DBG("failed reading status reg");
+		return;
+	}
+
+	if (status.drdy_pres != 0) {
```

Comment:
I can see some problems here when I tested an adapted version of this PR for an lps28dfw pressure sensor.

`s/if (status.drdy_pres != 0)/if (!status.drdy)`
`status.drdy_pres` should be `1` (status reg; P_DA bit) if pressure data is available.

The pin interrupt gets deactivated in `lps22df_gpio_callback` before this function is called. In case we have a spurious interrupt, we should probably configure the interrupt pin so that we will get future interrupts.

Reply:
> I can see some problems here when I tested an adapted version of this PR for an lps28dfw pressure sensor.
> 
> `s/if (status.drdy_pres != 0)/if (!status.drdy)` `status.drdy_pres` should be `1` (status reg; P_DA bit) if pressure data is available.
> 
> The pin interrupt gets deactivated in `lps22df_gpio_callback` before this function is called. In case we have a spurious interrupt, we should probably configure the interrupt pin so that we will get future interrupts.

Yes, sure. Good catch!

I may add an exit: strategy at the bottom:

```
        if (status.drdy_pres != 0) {
                goto exit; /* spurious interrupt */
        }

exit:
        ret = gpio_pin_interrupt_configure_dt(&cfg->gpio_int,
                                              GPIO_INT_EDGE_TO_ACTIVE);
        if (ret < 0) {
                LOG_ERR("%s: Not able to configure pin_int", dev->name);
        }

Reply:
done

Reply:
The exit strategy looks good, but the drdy_pres bit is still inverted? If drdy_pres is 1, data is available and therefore we do not have a spurious interrupt.

Reply:
> The exit strategy looks good, but the drdy_pres bit is still inverted? If drdy_pres is 1, data is available and therefore we do not have a spurious interrupt.

Yes sure... not sure how I can be so stupid. Thanks very much for being so patient...

Reply:
hope it might be a good starting point to extend the support to LPS28DFW:
https://github.com/zephyrproject-rtos/zephyr/issues/64766